test/src/test/java/org/apache/accumulo/test/functional/TableChangeStateIT.java

```diff
@@ -0,0 +1,549 @@
+import org.apache.accumulo.core.client.TableExistsException;
+import org.apache.accumulo.core.client.TableNotFoundException;
+import org.apache.accumulo.core.client.impl.Tables;
+import org.apache.accumulo.core.data.Key;
+import org.apache.accumulo.core.data.Mutation;
+import org.apache.accumulo.core.data.Value;
+import org.apache.accumulo.core.master.state.tables.TableState;
+import org.apache.accumulo.core.security.Authorizations;
+import org.apache.accumulo.core.zookeeper.ZooUtil;
+import org.apache.accumulo.fate.zookeeper.ZooCache;
+import org.apache.accumulo.fate.zookeeper.ZooReader;
+import org.apache.hadoop.io.Text;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * ACCUMULO-4574. Test to verify that changing table state to online / offline {@link org.apache.accumulo.core.client.admin.TableOperations#online(String)} when
+ * the table is already in that state returns without blocking.
+ */
+public class TableChangeStateIT extends ConfigurableMacIT {
+
+  private static final Logger log = LoggerFactory.getLogger(TableChangeStateIT.class);
+
+  private static final int NUM_ROWS = 1000;
+
+  @Override
+  protected int defaultTimeoutSeconds() {
+    return 4 * 60;
+  }
+
+  /**
+   * Validate that {@code TableOperations} online operation does not block when table is already online and fate transaction lock is held by other operations.
+   * The test creates, populates a table and then runs a compaction with a slow iterator so that operation takes long enough to simulate the condition. After
+   * the online operation while compaction is running completes, the test is complete and the compaction is canceled so that other tests can run.
+   *
+   * @throws Exception
+   *           any exception is a test failure.
+   */
+  @Test
+  public void changeTableStateTest() throws Exception {
+
+    Connector connector = getConnector();
+    String tableName = getUniqueNames(1)[0];
+
+    createData(tableName);
+
+    assertEquals("verify table online after created", TableState.ONLINE, getTableState(connector, tableName));
+
+    OnlineStatus status = onLine(tableName);
+
+    log.trace("Online 1 in {} ms", TimeUnit.MILLISECONDS.convert(status.runningTime(), TimeUnit.NANOSECONDS));
+
+    assertFalse("verify no exception thrown changing state", status.hasError());
+    assertEquals("verify table is still online", TableState.ONLINE, getTableState(connector, tableName));
+
+    // verify that offline then online functions as expected.
+
+    connector.tableOperations().offline(tableName, true);
```

Comment:
That's a good point.  The only FATE op I know of that works w/ an offline table is clone.